protocols/raft/src/main/java/io/atomix/protocols/raft/roles/LeaderRole.java

```diff
@@ -179,6 +186,106 @@ private void appendMembers() {
   }
 
   /**
+   * Starts checking for session heartbeat timeouts.
+   */
+  private void startHeartbeatTimer() {
+    raft.getSessions().getSessions().forEach(s -> s.resetHeartbeats());
+    log.trace("Starting heartbeat timers");
+    raft.getSessions().getSessions().stream()
+        .map(s -> s.memberId())
+        .distinct()
+        .forEach(this::resetHeartbeatTimer);
+  }
+
+  /**
+   * Resets the heartbeat timer.
+   */
+  private void resetHeartbeatTimer(MemberId member) {
+    // Compute the smallest timeout of all open sessions for the member.
+    OptionalLong minTimeout = raft.getSessions().getSessions().stream()
+        .filter(s -> s.memberId().equals(member))
+        .mapToLong(s -> s.minTimeout())
+        .min();
+
+    Scheduled oldTimer;
+    if (minTimeout.isPresent()) {
+      Scheduled newTimer = raft.getThreadContext().schedule(
```

Comment:
It sure if we should be using a recurring timer or setting a timer recursively when responses are received.